File: server/agents/overallAnalysisAgent.js
import { BaseAgent } from './baseAgent.js';

/**
 * Overall Analysis Agent
 * Synthesizes individual response analyses into comprehensive interview performance insights
 */
export class OverallAnalysisAgent extends BaseAgent {
  constructor(llmService) {
    const systemPrompt = `You are an Overall Analysis Agent that synthesizes interview performance data.

Your role is to:
1. Analyze patterns across all interview responses
2. Identify overall strengths and improvement areas
3. Calculate comprehensive performance metrics
4. Provide strategic recommendations for improvement
5. Generate executive summary of interview performance

Always respond with a JSON object containing:
{
  "overallScore": number (0-100),
  "performanceLevel": "excellent|good|fair|needs_improvement",
  "strengths": ["overall strength 1", "overall strength 2"],
  "improvements": ["strategic improvement 1", "strategic improvement 2"],
  "responseAnalysis": {
    "clarity": number (0-100),
    "structure": number (0-100),
    "technical": number (0-100),
    "communication": number (0-100),
    "confidence": number (0-100)
  },
  "trends": {
    "improvement": "improving|declining|consistent",
    "consistency": "high|medium|low",
    "adaptability": "high|medium|low"
  },
  "recommendations": ["recommendation 1", "recommendation 2"],
  "executiveSummary": "comprehensive summary paragraph",
  "nextSteps": ["next step 1", "next step 2"]
}

Provide strategic, actionable insights for interview improvement.`;

    super('OverallAnalysisAgent', llmService, systemPrompt);
  }

  preparePrompt(input, context) {
    const { responseAnalyses, config, sessionMetadata } = input;

    const analysisData = responseAnalyses.map((analysis, index) => ({
      questionNumber: index + 1,
      question: analysis.question,
      response: analysis.response,
      scores: analysis.analysis.responseAnalysis,
      overallScore: analysis.analysis.score,
      strengths: analysis.analysis.strengths,
      improvements: analysis.analysis.improvements
    }));

    return `Analyze the overall interview performance based on individual response analyses:

INTERVIEW CONTEXT:
- Topic: ${config.topic}
- Style: ${config.style}
- Experience Level: ${config.experienceLevel}
- Company: ${config.companyName || 'General'}
- Duration: ${config.duration} minutes
- Total Questions: ${responseAnalyses.length}

INDIVIDUAL RESPONSE ANALYSES:
${JSON.stringify(analysisData, null, 2)}

SESSION METADATA:
${JSON.stringify(sessionMetadata, null, 2)}

ANALYSIS REQUIREMENTS:
1. Calculate overall performance metrics across all responses
2. Identify patterns and trends in performance
3. Synthesize strengths that appear consistently
4. Identify improvement areas that need attention
5. Provide strategic recommendations for future interviews
6. Generate an executive summary of the candidate's performance

Focus on:
- Performance consistency across questions
- Improvement or decline trends during the interview
- Adaptability to different question types
- Overall readiness for the target role
- Specific, actionable next steps for improvement`;
  }

  processResponse(response, input, context) {
    try {
      // Clean the response to remove markdown code block delimiters
      let cleanedResponse = response.trim();
      
      console.log(`[OverallAnalysisAgent] Original response length: ${cleanedResponse.length}`);
      console.log(`[OverallAnalysisAgent] Response starts with: "${cleanedResponse.substring(0, 50)}..."`);
      
      // Remove leading ```json or ``` and trailing ```
      let cleaningApplied = false;
      
      if (cleanedResponse.startsWith('```json')) {
        cleanedResponse = cleanedResponse.substring(7);
        cleaningApplied = true;
        console.log('[OverallAnalysisAgent] Removed leading ```json delimiter');
      } else if (cleanedResponse.startsWith('```')) {
        cleanedResponse = cleanedResponse.substring(3);
        cleaningApplied = true;
        console.log('[OverallAnalysisAgent] Removed leading ``` delimiter');
      }
      
      if (cleanedResponse.endsWith('```')) {
        cleanedResponse = cleanedResponse.substring(0, cleanedResponse.length - 3);
        cleaningApplied = true;
        console.log('[OverallAnalysisAgent] Removed trailing ``` delimiter');
      }
      
      // Trim any remaining whitespace
      cleanedResponse = cleanedResponse.trim();
      
      console.log(`[OverallAnalysisAgent] Cleaned response length: ${cleanedResponse.length}`);
      console.log(`[OverallAnalysisAgent] Cleaning applied: ${cleaningApplied}`);
      console.log(`[OverallAnalysisAgent] Cleaned response starts with: "${cleanedResponse.substring(0, 50)}..."`);
      
      const result = JSON.parse(cleanedResponse);
      
      // Validate response structure
      if (!result.overallScore || !result.responseAnalysis) {
        throw new Error('Invalid overall analysis structure');
      }
      
      // Ensure scores are within valid range
      const scores = result.responseAnalysis;
      const scoreFields = ['clarity', 'structure', 'technical', 'communication', 'confidence'];
      
      for (const field of scoreFields) {
        if (typeof scores[field] !== 'number' || scores[field] < 0 || scores[field] > 100) {
          scores[field] = 70; // Default to neutral score
        }
      }
      
      // Validate overall score
      if (typeof result.overallScore !== 'number' || result.overallScore < 0 || result.overallScore > 100) {
        result.overallScore = Math.round(Object.values(scores).reduce((sum, score) => sum + score, 0) / scoreFields.length);
      }
      
      // Determine performance level if not provided
      if (!result.performanceLevel) {
        if (result.overallScore >= 85) result.performanceLevel = 'excellent';
        else if (result.overallScore >= 70) result.performanceLevel = 'good';
        else if (result.overallScore >= 60) result.performanceLevel = 'fair';
        else result.performanceLevel = 'needs_improvement';
      }
      
      console.log(`[OverallAnalysisAgent] Successfully parsed analysis with overall score: ${result.overallScore}`);
      
      return {
        success: true,
        analysis: result,
        metadata: {
          analyzedAt: new Date().toISOString(),
          totalResponses: input.responseAnalyses.length,
          averageResponseLength: input.responseAnalyses.reduce((sum, r) => sum + r.response.length, 0) / input.responseAnalyses.length,
          cleaningApplied
        }
      };
    } catch (error) {
      console.error('[OverallAnalysisAgent] Failed to parse overall analysis:', error);
      console.error('[OverallAnalysisAgent] Raw response preview:', response.substring(0, 300) + '...');
      
      // Try to extract JSON from the response if it's embedded in text
      try {
        console.log('[OverallAnalysisAgent] Attempting to extract JSON from response...');
        
        // Look for JSON object patterns in the response
        const jsonMatch = response.match(/\{[\s\S]*\}/);
        if (jsonMatch) {
          const extractedJson = jsonMatch[0];
          console.log(`[OverallAnalysisAgent] Found potential JSON, length: ${extractedJson.length}`);
          
          const result = JSON.parse(extractedJson);
          
          // Apply same validation as above
          if (result.overallScore && result.responseAnalysis) {
            console.log('[OverallAnalysisAgent] Successfully extracted and parsed JSON from response');
            
            // Ensure scores are within valid range
            const scores = result.responseAnalysis;
            const scoreFields = ['clarity', 'structure', 'technical', 'communication', 'confidence'];
            
            for (const field of scoreFields) {
              if (typeof scores[field] !== 'number' || scores[field] < 0 || scores[field] > 100) {
                scores[field] = 70;
              }
            }
            
            if (typeof result.overallScore !== 'number' || result.overallScore < 0 || result.overallScore > 100) {
              result.overallScore = Math.round(Object.values(scores).reduce((sum, score) => sum + score, 0) / scoreFields.length);
            }
            
            if (!result.performanceLevel) {
              if (result.overallScore >= 85) result.performanceLevel = 'excellent';
              else if (result.overallScore >= 70) result.performanceLevel = 'good';
              else if (result.overallScore >= 60) result.performanceLevel = 'fair';
              else result.performanceLevel = 'needs_improvement';
            }
            
            return {
              success: true,
              analysis: result,
              metadata: {
                analyzedAt: new Date().toISOString(),
                totalResponses: input.responseAnalyses.length,
                averageResponseLength: input.responseAnalyses.reduce((sum, r) => sum + r.response.length, 0) / input.responseAnalyses.length,
                extractedFromText: true
              }
            };
          }
        }
      } catch (extractError) {
        console.error('[OverallAnalysisAgent] JSON extraction also failed:', extractError);
      }
      
      // Generate fallback analysis
      console.log('[OverallAnalysisAgent] Using fallback analysis due to parsing failure');
      return {
        success: false,
        analysis: this.generateFallbackAnalysis(input),
        metadata: {
          analyzedAt: new Date().toISOString(),
          totalResponses: input.responseAnalyses.length,
          fallback: true,
          parseError: error.message,
          originalResponseLength: response.length
        }
      };
    }
  }

  generateFallbackAnalysis(input) {
    const { responseAnalyses, config } = input;
    
    console.log('[OverallAnalysisAgent] Generating fallback analysis');
    
    // Calculate averages from individual analyses
    const avgScores = {
      clarity: 0,
      structure: 0,
      technical: 0,
      communication: 0,
      confidence: 0
    };
    
    let totalScore = 0;
    
    responseAnalyses.forEach(analysis => {
      const scores = analysis.analysis.responseAnalysis;
      Object.keys(avgScores).forEach(key => {
        avgScores[key] += scores[key] || 70;
      });
      totalScore += analysis.analysis.score || 70;
    });
    
    Object.keys(avgScores).forEach(key => {
      avgScores[key] = Math.round(avgScores[key] / responseAnalyses.length);
    });
    
    const overallScore = Math.round(totalScore / responseAnalyses.length);
    
    let performanceLevel = 'fair';
    if (overallScore >= 85) performanceLevel = 'excellent';
    else if (overallScore >= 70) performanceLevel = 'good';
    else if (overallScore < 60) performanceLevel = 'needs_improvement';
    
    console.log(`[OverallAnalysisAgent] Fallback analysis generated with overall score: ${overallScore}`);
    
    return {
      overallScore,
      performanceLevel,
      strengths: ["Shows understanding of core concepts", "Demonstrates relevant experience"],
      improvements: ["Provide more specific examples", "Structure responses more clearly"],
      responseAnalysis: avgScores,
      trends: {
        improvement: "consistent",
        consistency: "medium",
        adaptability: "medium"
      },
      recommendations: [
        "Practice structuring responses using frameworks like STAR method",
        "Prepare specific examples for common question types",
        "Work on confident delivery and clear communication"
      ],
      executiveSummary: `The candidate demonstrated ${performanceLevel} performance with an overall score of ${overallScore}%. They show solid understanding of ${config.topic} concepts but could benefit from more structured responses and specific examples.`,
      nextSteps: [
        "Practice mock interviews focusing on response structure",
        "Prepare a portfolio of specific examples for different scenarios",
        "Work on confident delivery and clear articulation"
      ]
    };
  }
}

File: python_backend/agents/overall_analysis_agent.py
"""
Overall Analysis Agent using Pydantic AI
"""
from typing import Type, Dict, Any, List
from pydantic import BaseModel
from loguru import logger

from .base_agent import BaseAgent
from models.interview_models import InterviewConfig, ResponseAnalysis, PerformanceAnalytics

class OverallAnalysisAgent(BaseAgent[PerformanceAnalytics]):
    """Agent for overall interview performance analysis"""
    
    def get_result_type(self) -> Type[PerformanceAnalytics]:
        return PerformanceAnalytics
    
    def get_system_prompt(self) -> str:
        return """You are an Overall Analysis Agent that synthesizes interview performance data.

Your role is to:
1. Analyze patterns across all interview responses
2. Identify overall strengths and improvement areas
3. Calculate comprehensive performance metrics
4. Provide strategic recommendations for improvement
5. Generate executive summary of interview performance

Provide strategic, actionable insights for interview improvement. Focus on:
- Performance consistency across questions
- Improvement or decline trends during the interview
- Adaptability to different question types
- Overall readiness for the target role
- Specific, actionable next steps for improvement"""
    
    async def analyze_overall_performance(
        self,
        response_analyses: List[Dict[str, Any]],
        config: InterviewConfig,
        session_metadata: Dict[str, Any]
    ) -> PerformanceAnalytics:
        """Analyze overall interview performance"""
        input_data = {
            "response_analyses": response_analyses,
            "config": config.model_dump(),
            "session_metadata": session_metadata
        }
        
        context = {
            "analysis_type": "overall_performance",
            "total_responses": len(response_analyses),
            "interview_style": config.style
        }
        
        return await self.execute(input_data, context)
    
    def generate_fallback_result(self, input_data: Dict[str, Any], context: Dict[str, Any]) -> PerformanceAnalytics:
        """Generate fallback overall analysis"""
        response_analyses = input_data.get("response_analyses", [])
        config_data = input_data.get("config", {})
        
        logger.warning(f"[{self.name}] Using fallback overall analysis")
        
        if not response_analyses:
            # Default analysis when no responses
            return PerformanceAnalytics(
                overall_score=70,
                performance_level="fair",
                strengths=["Shows willingness to participate"],
                improvements=["Complete more interview questions"],
                response_analysis=ResponseAnalysis(
                    clarity=70, structure=70, technical=70,
                    communication=70, confidence=70, relevance=70
                ),
                trends={"improvement": "consistent", "consistency": "medium", "adaptability": "medium"},
                recommendations=["Practice more interview scenarios"],
                executive_summary="Limited data available for comprehensive analysis.",
                next_steps=["Complete full interview sessions"],
                question_reviews=[],
                metadata={"fallback": True, "total_responses": 0}
            )
        
        # Calculate averages from individual analyses
        avg_scores = {
            "clarity": 0, "structure": 0, "technical": 0,
            "communication": 0, "confidence": 0, "relevance": 0
        }
        
        total_score = 0
        all_strengths = []
        all_improvements = []
        question_reviews = []
        
        for i, analysis in enumerate(response_analyses):
            analysis_data = analysis.get("analysis", {})
            scores = analysis_data.get("response_analysis", {})
            
            # Accumulate scores
            for key in avg_scores:
                avg_scores[key] += scores.get(key, 70)
            
            total_score += analysis_data.get("score", 70)
            
            # Collect strengths and improvements
            if analysis_data.get("strengths"):
                all_strengths.extend(analysis_data["strengths"])
            if analysis_data.get("improvements"):
                all_improvements.extend(analysis_data["improvements"])
            
            # Create question review
            question_reviews.append({
                "question_id": analysis.get("question_id", f"q{i+1}"),
                "question": analysis.get("question", ""),
                "response": analysis.get("response", ""),
                "score": analysis_data.get("score", 70),
                "feedback": analysis_data.get("feedback", "Good response")
            })
        
        # Calculate averages
        num_responses = len(response_analyses)
        for key in avg_scores:
            avg_scores[key] = round(avg_scores[key] / num_responses)
        
        overall_score = round(total_score / num_responses)
        
        # Determine performance level
        if overall_score >= 85:
            performance_level = "excellent"
        elif overall_score >= 70:
            performance_level = "good"
        elif overall_score >= 60:
            performance_level = "fair"
        else:
            performance_level = "needs_improvement"
        
        # Deduplicate strengths and improvements
        unique_strengths = list(set(all_strengths))[:3] if all_strengths else [
            "Shows understanding of core concepts",
            "Demonstrates relevant experience",
            "Communicates ideas clearly"
        ]
        
        unique_improvements = list(set(all_improvements))[:3] if all_improvements else [
            "Provide more specific examples",
            "Structure responses more clearly",
            "Practice confident delivery"
        ]
        
        return PerformanceAnalytics(
            overall_score=overall_score,
            performance_level=performance_level,
            strengths=unique_strengths,
            improvements=unique_improvements,
            response_analysis=ResponseAnalysis(**avg_scores),
            trends={
                "improvement": "consistent",
                "consistency": "medium",
                "adaptability": "medium"
            },
            recommendations=[
                "Practice structuring responses using frameworks like STAR method",
                "Prepare specific examples for common question types",
                "Work on confident delivery and clear communication",
                f"Focus on improving {config_data.get('topic', 'technical')} knowledge depth"
            ],
            executive_summary=f"The candidate demonstrated {performance_level} performance with an overall score of {overall_score}%. They show solid understanding of {config_data.get('topic', 'the subject')} concepts but could benefit from more structured responses and specific examples.",
            next_steps=[
                "Practice mock interviews focusing on response structure",
                "Prepare a portfolio of specific examples for different scenarios",
                "Work on confident delivery and clear articulation",
                f"Deepen knowledge in {config_data.get('topic', 'the subject')} through additional study and practice"
            ],
            question_reviews=question_reviews,
            metadata={
                "fallback": True,
                "total_responses": num_responses,
                "analysis_method": "fallback"
            }
        )